src/network/networkd-dhcp4.c

```diff
@@ -138,6 +139,151 @@ static bool link_prefixroute(Link *link) {
+        }
+
+        return 0;
+}
+
+static int dhcp_routing_policy_rule_compare_func(union in_addr_union *a, union in_addr_union *b, unsigned rule_prefixlen, unsigned dhcp_prefixlen) {
+        int r;
+
+        assert(a);
+        assert(b);
+
+        r = CMP(rule_prefixlen, dhcp_prefixlen);
+        if (r != 0)
+                return r;
+
+        return memcmp(a, b, FAMILY_ADDRESS_SIZE(AF_INET));
+}
+
+static int dhcp_remove_routing_policy_rules(Link *link) {
+        const struct in_addr *router = NULL;
+        union in_addr_union address;
+        RoutingPolicyRule *rule;
+        struct in_addr netmask;
+        unsigned prefixlen;
+        int k, r;
+
+        assert(link);
+        assert(link->network);
+        assert(link->manager);
+        assert(link->dhcp_lease);
+
+        r = sd_dhcp_lease_get_address(link->dhcp_lease, &address.in);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no address: %m");
+
+        r = sd_dhcp_lease_get_netmask(link->dhcp_lease, &netmask);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no netmask: %m");
+
+        prefixlen = in4_addr_netmask_to_prefixlen(&netmask);
+
+        (void) sd_dhcp_lease_get_router(link->dhcp_lease, &router);
+
+        if (DEBUG_LOGGING) {
+                _cleanup_free_ char *pretty = NULL;
+
+                (void) in_addr_to_string(AF_INET, &address, &pretty);
+                log_link_debug(link, "DHCP: removing RPDB rule from DHCPv4 address %s", strna(pretty));
+        }
+
+        HASHMAP_FOREACH(rule, link->network->rules_by_section) {
+                union in_addr_union rt = {
+                        .in = *router,
+                };
+
+                if (!IN_SET(rule->to_type, ADDRESS_TYPE_DHCP4_IP, ADDRESS_TYPE_DHCP4_GATEWAY) &&
+                    !IN_SET(rule->from_type, ADDRESS_TYPE_DHCP4_IP, ADDRESS_TYPE_DHCP4_GATEWAY))
+                        continue;
+
+                if ((rule->to_type == ADDRESS_TYPE_DHCP4_GATEWAY || rule->from_type == ADDRESS_TYPE_DHCP4_GATEWAY) && !router)
+                        continue;
+
+                if (rule->from_type == ADDRESS_TYPE_DHCP4_IP && (dhcp_routing_policy_rule_compare_func(&rule->from, &address, rule->from_prefixlen, prefixlen) != 0))
+                        continue;
+
+                if (rule->to_type == ADDRESS_TYPE_DHCP4_IP && (dhcp_routing_policy_rule_compare_func(&rule->to, &address, rule->to_prefixlen, prefixlen) != 0))
+                        continue;
+
+                if (rule->from_type == ADDRESS_TYPE_DHCP4_GATEWAY && (dhcp_routing_policy_rule_compare_func(&rule->from, &rt, rule->from_prefixlen, 32) != 0))
+                        continue;
+
+                if (rule->to_type == ADDRESS_TYPE_DHCP4_GATEWAY && (dhcp_routing_policy_rule_compare_func(&rule->to, &rt, rule->to_prefixlen, 32) != 0))
+                        continue;
```

Comment:
No no, please re-assign `from` or `to`, as similar to `dhcp_set_routing_policy_rules()` in the above.

Please consider the following case,
```
[Match]
Name=eth*

DHCP=ipv4

[RoutingPolicyRule]
From=_dhcp4_ip
To=_dhcp4_gw
```
If this .network file matches multiple interfaces, say eth0 and eth1, and these interfaces connected to different network, and each network has different DHCPv4 server. 
- First, e.g. eth0 acquires a DHCP lease and edits RoutingPolicyRule object in the corresponding Network object, 
- Next, eth1 acquires a DHCP lease and also edits the same RoutingPolicyRule according the lease for eth1.
- When the lease for eth0 expires, then the above logic does not match, as the object was edited based on the lease for eth1.

Reply:
Hmm, after re-considering the removing logic. How about to take a reference of rules created in `dhcp_set_routing_policy_rules()`? Like `dhcp_route_configure()`,

Reply:
>If this .network file matches multiple interfaces, say eth0 and eth1, and these interfaces connected to different network, and each network has different DHCPv4 server.

Then both have will have different IP and GW. 

>First, e.g. eth0 acquires a DHCP lease and edits RoutingPolicyRule object in the corresponding Network object,

A new rule will be allocated and inserted according to the `From=` and `To=`

>Next, eth1 acquires a DHCP lease and also edits the same RoutingPolicyRule according the lease for eth1.

Another new rule will be allocated and inserted according to the `From=`  and `To=`


```

ip rule
0:	from all lookup local
112:	from 172.16.85.225/24 to 172.16.85.225/24 tos 0x10 lookup 10 proto static
112:	from 172.16.85.131/24 to 172.16.85.131/24 tos 0x10 lookup 10 proto static
32766:	from all lookup main
32767:	from all lookup default

```
>When the lease for eth0 expires, then the above logic does not match, as the object was edited based on the lease for eth1.

No when eth0 expires the rules which is allocated according to the settings will be removed. eth1 still have own rule.
We are pulling out the data and comparing if this rule object is associated with the lease. 

Note that we want to remove the rule which we added. delete only from and to removes another one and from this rule should be also be deleted in userspace
```
 ❯ sudo ip rule add from 192.186.100.176/24 table 11
                                                                                                                                                                       
 ❯ sudo ip rule add from 192.186.100.176/24 table 12
                                                                                                                                                                       
 ❯ ip rule
0:	from all lookup local
32763:	from 192.186.100.176/24 lookup 12
32764:	from 192.186.100.176/24 lookup 11
32765:	from 192.186.100.176/24 lookup main


 ❯ ip rule                                 
0:	from all lookup local
32761:	from 192.186.100.176/24 lookup main
32762:	from 192.186.100.176/24 lookup main
32763:	from 192.186.100.176/24 lookup 12
32764:	from 192.186.100.176/24 lookup 11
32765:	from 192.186.100.176/24 lookup main
32766:	from all lookup main
32767:	from all lookup default
                                                                                                                                                                       
 ❯ sudo ip rule delete from 192.186.100.176/24 
                                                                                                                                                                       
 ❯ ip rule                                    
0:	from all lookup local
32762:	from 192.186.100.176/24 lookup main
32763:	from 192.186.100.176/24 lookup 12
32764:	from 192.186.100.176/24 lookup 11
32765:	from 192.186.100.176/24 lookup main
32766:	from all lookup main
32767:	from all lookup default
                                                                                                                                                                       
❯ sudo ip rule delete from 192.186.100.176/24
                                                                                                                                                                       
 ❯ ip rule                                    
0:	from all lookup local
32763:	from 192.186.100.176/24 lookup 12
32764:	from 192.186.100.176/24 lookup 11
32765:	from 192.186.100.176/24 lookup main
32766:	from all lookup main
32767:	from all lookup default
                                                                                                                                                                       
 ❯ sudo ip rule delete from 192.186.100.176/24
                                                                                                                                                                       
 ❯ ip rule                                    
0:	from all lookup local
32764:	from 192.186.100.176/24 lookup 11
32765:	from 192.186.100.176/24 lookup main
32766:	from all lookup main
32767:	from all lookup default
                                           
 
sudo ip rule delete from 192.186.100.176/24
                                                                                                                                                                       
❯ ip rule                                    
0:	from all lookup local
32765:	from 192.186.100.176/24 lookup main
32766:	from all lookup main
32767:	from all lookup default
❯ sudo ip rule delete from 192.186.100.176/24
                                                                                                                                                                       
 ❯ ip rule                                    
0:	from all lookup local
32766:	from all lookup main
32767:	from all lookup default
```


and I tested 
```
ip rule
0:	from all lookup local
112:	from 172.16.85.225/24 to 172.16.85.225/24 tos 0x10 lookup 10 proto static
112:	from 172.16.85.131/24 to 172.16.85.131/24 tos 0x10 lookup 10 proto static
32766:	from all lookup main
32767:	from all lookup default

```

Reply:
> Hmm, after re-considering the removing logic. How about to take a reference of rules created in `dhcp_set_routing_policy_rules()`? Like `dhcp_route_configure()`,

The only case where this makes sense where both interfaces connected to same DHCP servers. Usually bonding is used for such option. We can add this logic in a follow up

Reply:
ping

Reply:
Sorry, for late response. But I cannot follow what you want to say... Please elaborate more, or implement a way you think it is correct. I hope new codes can explain well what you want to say, and maybe uncover further problems.

Reply:
It's very simple and strate forward. If you test then you will figure this out. 2 GW means 2 rules since they are dynamic.